[core] Handle std::optional spell OnMobMagicPrepare

sruon opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Resolves a crash when the initial spellId is nullopt. Seen in a custom CE script.

Exact steps to reproduce elude me right now, but the cause was fairly obvious.

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->

sruon committed:
Handle optional SpellID

Co-Authored-By: Zach Toogood <1389729+zach2good@users.noreply.github.com>

zach2good commented:
Putting a hold on this so it doesn't get merged before  I can take a look - at a glance I'm not sure about this logic

zach2good commented:
This is the logic that leads up to this:
```cpp
bool CMobController::TryCastSpell()
{
    TracyZoneScoped;
    if (!CanCastSpells())
    {
        return false;
    }

    // Find random spell from list
    std::optional<SpellID> chosenSpellId;
    if (m_firstSpell)
    {
        // mobs first spell, should be aggro spell
        chosenSpellId = PMob->SpellContainer->GetAggroSpell();
        m_firstSpell  = false;
    }
    else
    {
        chosenSpellId = PMob->SpellContainer->GetSpell();
    }

    // Try to get an override spell from the script (if available)
    auto PSpellTarget            = PTarget ? PTarget : PMob;
    auto possibleOverriddenSpell = luautils::OnMobMagicPrepare(PMob, PSpellTarget, chosenSpellId);
    if (possibleOverriddenSpell.has_value())
    {
        chosenSpellId = possibleOverriddenSpell;
    }

    if (chosenSpellId.has_value())
    {
        CastSpell(chosenSpellId.value());
        return true;
    }

    return false;
}
```

This PR smells very much like it's sweeping the symptoms of the bug under the rug, rather than fixing the actual bug. 

This is the block that needs inspecting:
```cpp
    // Find random spell from list
    std::optional<SpellID> chosenSpellId;
    if (m_firstSpell)
    {
        // mobs first spell, should be aggro spell
        chosenSpellId = PMob->SpellContainer->GetAggroSpell();
        m_firstSpell  = false;
    }
    else
    {
        chosenSpellId = PMob->SpellContainer->GetSpell();
    }
```

Should the `chosenSpellId` be optional at all?

zach2good commented on src/map/lua/luautils.cpp:
```diff
@@ -2845,7 +2845,7 @@ namespace luautils
         }
 
-        auto result = onMobMagicPrepare(PCaster, PTarget, spell::GetSpell(startingSpellId.value()));
+        auto result = onMobMagicPrepare(PCaster, PTarget, startingSpellId.has_value() ? spell::GetSpell(startingSpellId.value()) : nullptr);
```
```cpp
CSpell* PSpell = nullptr;
if (startingSpellId.has_value())
{
    PSpell = spell::GetSpell(startingSpellId.value());
}

auto result = onMobMagicPrepare(PCaster, PTarget, PSpell);
```

sruon commented:
https://github.com/LandSandBoat/server/blob/162066bf564a9f24e44418cd0d87c75933ce805d/src/map/mob_spell_container.cpp#L844-L852

For record sake, this is what led to the crash. 

Fomor (with a Cure spell in spell list) did not pass this check, leading to nullopt being fed to `GetSpell`.

sruon commented:
I updated the proposal with zach changes, and also removed the undead check since Fomors legitimately cast Cure on themselves on retail.

![Screenshot 2025-04-18 130503](https://github.com/user-attachments/assets/b7eca6f2-4d6f-4701-804d-985f9d90a1f4)

[Capture](https://drive.google.com/file/d/1XRuHUx_2udhLRpI0XOd_T26b2V3AnIDE/view?usp=sharing)

List of mobs that will be impacted by this change and may require spell lists adjustments in the future:

- Caithleann
- Fomor_Paladin
- Fomor_Red_Mage
- Giollemitte_B_Feroun
- Helmwige
- Hydra_Paladin
- Hydra_Red_Mage
- Hydra_White_Mage
- Ortlinde
- Powderkeg_Yanadahn
- Southern_Spriggan
- Indomitable_Spurned
- Ephramadian_Shade_rdm
- Shunned_Deeparrow
- Shunned_Nightstalker
- Shunned_Hexer
- Shunned_Saboteur
- Shunned_Fluteslinger
- Shunned_Elementalist
- Shunned_Engraver
- Shunned_Valiant
- Vengeful_Shunned_1
- Ashu_Talif_Crew_rdm
- Spurned_Saboteur
- Spurned_Fluteslinger
- Spurned_Nightstalker
- Spurned_Deeparrow
- Spurned_Elementalist
- Spurned_Engraver
- Spurned_Valiant
- Spurned_Hexer
- Indomitable_Spurned_1
- Indomitable_Spurned_2
- Indomitable_Spurned_3
- Indomitable_Spurned_4
- Indomitable_Spurned_5
- Experimental_Undead_blm
- Pignonpausard
- Fir_Bholg_pld_gk
- Fir_Bholg_rdm_mr
- Hydra_Paladin_HI
- Hydra_Red_Mage_HI
- Hydra_White_Mage_HI
- Fomor_Paladin_LM
- Fomor_Paladin_WW_BT
- Fomor_Red_Mage_LM
- Shadowbreath
- Australis_Shadow

```
SELECT DISTINCT
    mp.name,
    mp.familyid,
    mp.spellList
FROM 
    mob_pools mp
JOIN 
    mob_family_system mfs ON mp.familyid = mfs.familyID
JOIN 
    mob_spell_lists msl ON mp.spellList = msl.spell_list_id
JOIN 
    spell_list sl ON msl.spell_id = sl.spellid
WHERE 
    mfs.ecosystem = 'Undead'
    AND sl.name LIKE 'Cure%';
```
